### PR TITLE
Som/Null Backtest fold - best model fix and Foundational Model Retries

### PR DESCRIPTION
**Pull request by smdp2000ofc**
This pull request introduces several robustness and reliability improvements to the forecasting pipeline, focusing on handling partial model results, improving error handling and retries for API calls, and enhancing utility/test coverage. The changes ensure that only models with complete back test coverage are eligible for selection as the "Best Model," add exponential backoff for transient API failures, and improve utility function correctness and testing.

**Model selection and back test coverage:**
* The `final_models` function now checks that only models with complete back test fold coverage are eligible for "Best Model" selection. Models missing any back test folds are flagged and excluded from selection, but their partial results remain in the output for visibility. If no models have complete coverage, an error is raised. [[1]](diffhunk://#diff-c261549c121fdfb143eebe844ce0c8f1bf21c76de5d9cf1bca5285b93ea81d2cR339-R415) [[2]](diffhunk://#diff-c261549c121fdfb143eebe844ce0c8f1bf21c76de5d9cf1bca5285b93ea81d2cR424-R438)
* The logic for building the allow-list of eligible models for "Best Model" selection is updated to include only complete models and valid averages. The output contract is preserved so every model row has a non-NA `Best_Model` flag. [[1]](diffhunk://#diff-c261549c121fdfb143eebe844ce0c8f1bf21c76de5d9cf1bca5285b93ea81d2cR548-R557) [[2]](diffhunk://#diff-c261549c121fdfb143eebe844ce0c8f1bf21c76de5d9cf1bca5285b93ea81d2cL586-R703)

**Chronos API reliability:**
* The `send_chronos_request` function now retries on transient failures (HTTP 429, 5xx, or connection errors) using exponential backoff with jitter, up to a maximum number of attempts. This makes the system more resilient to temporary API issues. [[1]](diffhunk://#diff-72a39c0970d6b42a7880b322ea144d31f5c434a9196a9ad704e9dd6183907566R151-R164) [[2]](diffhunk://#diff-72a39c0970d6b42a7880b322ea144d31f5c434a9196a9ad704e9dd6183907566L166-R245)

**Utility and testing improvements:**
* The `null_converter` utility function is improved to handle more edge cases (NULL, zero-length, "NULL" string, NA) and now passes through non-NULL values and vectors of length > 1. Comprehensive tests are added to ensure correct behavior. [[1]](diffhunk://#diff-0c8a2fab8eac2442c77c71d71f1ecc55882b5a4ab0144f58fbd8d02ff42a3272R2485-R2494) [[2]](diffhunk://#diff-449f8a7b589082de8669172853b0c62097a2ccb6accb14b6da7397e95473fb14R1116-R1153)
* Additional variables are registered as global to avoid R CMD check warnings.

**Documentation and namespace clean-up:**
* The S3 method registration for `print.timesfm_model_fit_impl` is removed from the namespace, and its documentation file is deleted, reflecting deprecation or removal of this method. [[1]](diffhunk://#diff-b6b6854a02ae177f8860f49654ff250c1554d021ae434b6efdbe99d00bdc6055L24) [[2]](diffhunk://#diff-b5a58007622f09132b7f7157b430b8ae98dca8d9b153814a343af14e6f2b3bb8L1-L20)This pull request introduces several robustness and reliability improvements to the forecasting pipeline, particularly around handling incomplete model evaluations and improving resilience to external API failures. The most significant changes are enhanced error handling and retry logic for Chronos API requests, stricter enforcement of model selection criteria, and improved utility/test coverage for helper functions.

**Reliability and Robustness Improvements:**

* Added exponential backoff and retry logic (up to 3 retries) to the `send_chronos_request` function to handle transient Chronos API failures (HTTP 429, 5xx, and connection errors), improving reliability of external API calls. [[1]](diffhunk://#diff-72a39c0970d6b42a7880b322ea144d31f5c434a9196a9ad704e9dd6183907566R151-R164) [[2]](diffhunk://#diff-72a39c0970d6b42a7880b322ea144d31f5c434a9196a9ad704e9dd6183907566L166-R245)
* Implemented stricter checks in `final_models` to ensure that only models with complete back test fold coverage are eligible for Best_Model selection, preventing partial models from being unfairly selected. Models with incomplete coverage are flagged and excluded from selection, but still included in outputs for transparency. [[1]](diffhunk://#diff-c261549c121fdfb143eebe844ce0c8f1bf21c76de5d9cf1bca5285b93ea81d2cR339-R415) [[2]](diffhunk://#diff-c261549c121fdfb143eebe844ce0c8f1bf21c76de5d9cf1bca5285b93ea81d2cR424-R438) [[3]](diffhunk://#diff-c261549c121fdfb143eebe844ce0c8f1bf21c76de5d9cf1bca5285b93ea81d2cR548-R557)
* Enhanced the Best_Model output logic so that every model in the output (including partial-coverage models) is flagged as either "Yes" or "No" for Best_Model, maintaining contract consistency.

**Utility and Testing Enhancements:**

* Improved the `null_converter` utility function to robustly handle `NULL`, zero-length, `"NULL"` string, and `NA` inputs, with comprehensive new unit tests added for all edge cases. [[1]](diffhunk://#diff-0c8a2fab8eac2442c77c71d71f1ecc55882b5a4ab0144f58fbd8d02ff42a3272R2485-R2494) [[2]](diffhunk://#diff-449f8a7b589082de8669172853b0c62097a2ccb6accb14b6da7397e95473fb14R1116-R1153)
* Updated `utils::globalVariables` to include new variables used in back test coverage tracking, preventing unnecessary warnings.

**Documentation and Namespace Clean-up:**

* Removed the S3 method registration and documentation for `print.timesfm_model_fit_impl`, reflecting its deprecation or removal. [[1]](diffhunk://#diff-b6b6854a02ae177f8860f49654ff250c1554d021ae434b6efdbe99d00bdc6055L24) [[2]](diffhunk://#diff-b5a58007622f09132b7f7157b430b8ae98dca8d9b153814a343af14e6f2b3bb8L1-L20)

These changes collectively harden the pipeline against transient failures, enforce fair model evaluation, and improve code maintainability and test coverage.This pull request introduces multiple improvements to model selection robustness, error handling, and testing in the forecasting pipeline. The main focus is to ensure that only models with complete back test coverage are eligible to win Best_Model, preventing models with missing folds (often due to transient API failures) from being unfairly selected. In addition, the Chronos API client now has robust retry logic, and comprehensive tests have been added to verify the new behavior.

**Model selection robustness and fairness:**

* In `final_models()`, models with incomplete back test fold coverage are now identified and excluded from Best_Model selection. Partial models are flagged with warnings, and only models with predictions for all folds are eligible. If no model has complete coverage, an error is raised. This prevents models that failed on some folds from winning due to being evaluated on a smaller, easier sample. [[1]](diffhunk://#diff-c261549c121fdfb143eebe844ce0c8f1bf21c76de5d9cf1bca5285b93ea81d2cR339-R383) [[2]](diffhunk://#diff-c261549c121fdfb143eebe844ce0c8f1bf21c76de5d9cf1bca5285b93ea81d2cR392-R406) [[3]](diffhunk://#diff-c261549c121fdfb143eebe844ce0c8f1bf21c76de5d9cf1bca5285b93ea81d2cR516-R525)
* The output now ensures every model row (including partials) is present, but only complete models can be flagged as Best_Model, maintaining full visibility and backward compatibility.

**Chronos API reliability:**

* The Chronos API client (`send_chronos_request`) now implements exponential backoff with jitter for transient failures (HTTP 429, 5xx, or connection errors), retrying up to a configurable maximum. This greatly improves robustness to intermittent API issues. [[1]](diffhunk://#diff-72a39c0970d6b42a7880b322ea144d31f5c434a9196a9ad704e9dd6183907566R151-R160) [[2]](diffhunk://#diff-72a39c0970d6b42a7880b322ea144d31f5c434a9196a9ad704e9dd6183907566L166-R241)

**Testing and utility updates:**

* Extensive tests were added to verify that models with partial back test coverage cannot win Best_Model, that errors are raised if no model has full coverage, and that selection metrics are computed only from complete models.
* Utility and documentation updates: Added `fold_count` to `globalVariables`, removed the unused `print.timesfm_model_fit_impl` alias and documentation, and improved the `null_converter` function for robustness. [[1]](diffhunk://#diff-777c2db7dd77ffcdb2fe2752209c94ac2aa919f31009d985c2642c879ad58ef9L31-R31) [[2]](diffhunk://#diff-b6b6854a02ae177f8860f49654ff250c1554d021ae434b6efdbe99d00bdc6055L24) [[3]](diffhunk://#diff-b5a58007622f09132b7f7157b430b8ae98dca8d9b153814a343af14e6f2b3bb8L1-L20) [[4]](diffhunk://#diff-0c8a2fab8eac2442c77c71d71f1ecc55882b5a4ab0144f58fbd8d02ff42a3272R2485-R2494)

These changes collectively ensure fairer model evaluation, more robust API interactions, and improved test coverage for critical selection logic.
This pull request introduces robust handling for transient API failures, improves the fairness and reliability of model selection, and strengthens utility/test coverage. The main changes are grouped below.

**Resilient API Request Handling:**
* Added exponential backoff with jitter and retry logic to both `send_chronos_request` and `send_timesfm_request`, retrying up to three times on transient errors (HTTP 429, 5xx, or connection issues) to improve reliability when communicating with Chronos and TimesFM APIs. [[1]](diffhunk://#diff-72a39c0970d6b42a7880b322ea144d31f5c434a9196a9ad704e9dd6183907566R151-R164) [[2]](diffhunk://#diff-72a39c0970d6b42a7880b322ea144d31f5c434a9196a9ad704e9dd6183907566L166-R243) [[3]](diffhunk://#diff-dcd6f7c77e2125b2bd04063427826277431ac62ac38a2ae6a69677945791dec2R332-R344) [[4]](diffhunk://#diff-dcd6f7c77e2125b2bd04063427826277431ac62ac38a2ae6a69677945791dec2L344-R422)

**Fair Model Selection and Evaluation:**
* In `final_models`, models that do not produce predictions for all expected back test folds are now detected and excluded from Best_Model selection, with detailed warnings and partial results retained for transparency. This prevents models with incomplete evaluations from being unfairly favored. [[1]](diffhunk://#diff-c261549c121fdfb143eebe844ce0c8f1bf21c76de5d9cf1bca5285b93ea81d2cR339-R415) [[2]](diffhunk://#diff-c261549c121fdfb143eebe844ce0c8f1bf21c76de5d9cf1bca5285b93ea81d2cR424-R438)
* Ensured that only eligible models (those with complete back test coverage and valid averages) are considered for Best_Model selection, and that all models (including partials) are reported in outputs with appropriate flags. [[1]](diffhunk://#diff-c261549c121fdfb143eebe844ce0c8f1bf21c76de5d9cf1bca5285b93ea81d2cR548-R557) [[2]](diffhunk://#diff-c261549c121fdfb143eebe844ce0c8f1bf21c76de5d9cf1bca5285b93ea81d2cL586-R703)

**Utility and Test Improvements:**
* Improved the `null_converter` utility to robustly handle a wider range of null-like inputs, and added comprehensive unit tests to verify its behavior. [[1]](diffhunk://#diff-0c8a2fab8eac2442c77c71d71f1ecc55882b5a4ab0144f58fbd8d02ff42a3272R2485-R2494) [[2]](diffhunk://#diff-449f8a7b589082de8669172853b0c62097a2ccb6accb14b6da7397e95473fb14R1116-R1153)
* Registered new global variables to avoid R CMD check warnings for new columns used in model coverage tracking.

**Documentation and Cleanup:**
* Removed the print method for `timesfm_model_fit_impl` from the NAMESPACE and deleted its documentation, as it is no longer needed. [[1]](diffhunk://#diff-b6b6854a02ae177f8860f49654ff250c1554d021ae434b6efdbe99d00bdc6055L24) [[2]](diffhunk://#diff-b5a58007622f09132b7f7157b430b8ae98dca8d9b153814a343af14e6f2b3bb8L1-L20)

These changes collectively make the modeling pipeline more robust, transparent, and maintainable.

### CODE REVIEW

File: R/chronos_controller.R [hunk 1]
```diff
@@ -163,17 +169,74 @@ send_chronos_request <- function(payload) {
...
+          sprintf(
+            "Chronos API request failed after %d attempts. Last error: %s",
+            attempt, conditionMessage(response)
+          ),
+          call. = FALSE
+        )
+      }
+      backoff_wait(attempt)
+      next
+    }
```
[the remark this answers is not in the document]
Comment (smdp2000ofc) [reply R/chronos_controller.R:208]: i think cap total retries at max 3 and total attempst 4 including the original call

**Comment by smdp2000ofc** [comment]
@copilot can you raise another PR for the suggestion you gave and also, look into the comments that i have for some suggestion

**Comment by smdp2000ofc** [comment]
@copilot can u review just last commit -
4153a1292b66622e510fcaa89a764d0ab083ebf4

File: R/final_models.R [hunk 2]
```diff
@@ -336,11 +336,83 @@ final_models <- function(run_info,
...
+        # folds would otherwise be evaluated on a smaller, often easier sample
+        # than its competitors and could win Best_Model unfairly. flag these
+        # models so downstream Best_Model selection excludes them, while still
+        # keeping their partial rows in the output for visibility. only back
+        # test Train_Test_ID values are considered here; future forecast output
+        # is not part of back test fold coverage.
+        back_test_id_list <- setdiff(train_test_id_list, 1)
+        expected_back_test_fold_count <- length(back_test_id_list)
+
+        if (expected_back_test_fold_count > 0) {
```
Comment (mitokic) [inline R/final_models.R:351]: Is there any reason why there wouldn't be back test folds?

File: R/final_models.R [hunk 3]
```diff
@@ -336,11 +336,83 @@ final_models <- function(run_info,
...
+            partial_model_id <- partial_models_tbl$Model_ID[i]
+            actual_ids <- partial_models_tbl$actual_ids[[i]]
+            missing_ids <- partial_models_tbl$missing_ids[[i]]
+            cli::cli_alert_warning(
+              "Combo '{partial_combo}': model '{partial_model_id}' produced only {length(actual_ids)} of {expected_back_test_fold_count} back test folds (missing fold IDs: {paste(missing_ids, collapse = ', ')}). Excluding from Best_Model selection."
+            )
+          }
+        }
+
+        complete_model_ids <- if (expected_back_test_fold_count > 0) {
```
Comment (mitokic) [inline R/final_models.R:398]: again why check for no back test folds. In theory there should always be a back test

**Review by mitokic — CHANGES_REQUESTED** [review]

**Review by mitokic — APPROVED** [review]
